### PR TITLE
Fix [`needless_return`] false negative 

### LINKED ISSUE #12907: Removing return not sugggested by `needless_return` when calling some chained string methods
### Summary

When having a return statement with a call to `str::trim`, the `needless_return` list does not warn. Replacing `trim` with`len` in the following example correctly warns.

It seems to happen with other borrowing methods as well, such as in the following example:
```rust
fn f() -> String {
	return "".split("").next().unwrap().to_string();
}
```

### Lint Name

needless_return

### Reproducer

I tried this code:

```rust
fn main(){
	f();
}
fn f() -> String {
	return "".trim().to_string();
}
```

I expected to see this happen:
```
warning: unneeded `return` statement
 --> src/main.rs:5:2
  |
5 |     return "".trim().to_string();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return
  = note: `#[warn(clippy::needless_return)]` on by default
help: remove `return`
  |
5 -     return "".trim().to_string();
5 +     "".trim().to_string()
  |
```

Instead, this happened:

No output

### Version

```text
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-unknown-linux-gnu
release: 1.78.0
LLVM version: 18.1.2
```

### PR DESCRIPTION
**Pull request by apoisternex**
Fixes #12907

changelog: Fix [`needless_return`] false negative when returned expression borrows a value.

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! Let's just wait for the assigned reviewer to take a look before approving.

File: tests/ui/needless_return.rs [hunk 1]
```diff
@@ -365,4 +365,8 @@ fn conjunctive_blocks() -> String {
     return { "a".to_string() } + "b" + { "c" };
 }
 
+fn issue12907() -> String {
+    return "".split("").next().unwrap().to_string();
+}
+
```
Comment (Centri3) [inline tests/ui/needless_return.rs:371]: I think this just needs a test for the mentioned `RefCell` or any other type that fails to compile. but I think the significant drop check is adequate

There is the `rustc_insigificant_dtor` attribute. (I didn't even know this existed before checking.) This is given on the types `Rc`, `HashMap`, and `array::IntoIter`. Not sure whether this attribute changes the behavior (i.e., compiling would succeed) but it'd be a good idea to check.

Comment (apoisternex) [reply tests/ui/needless_return.rs:371]: > I think this just needs a test for the mentioned `RefCell` or any other type that fails to compile. But I think the significant drop check is adequate

So should i delete the test then? 

> Not sure whether this attribute changes the behavior (i.e., compiling would succeed) but it'd be a good idea to check.

Yes, it compiles with the rustc_insigificant_dtor attribute and the lint is triggered.

Comment (Centri3) [reply tests/ui/needless_return.rs:371]: No I meant the only change it needs

Comment (apoisternex) [reply tests/ui/needless_return.rs:371]: There is already a very similar test to the one mentioned:

``` rust
fn temporary_outlives_local() -> String {
    let x = RefCell::<String>::default();
    return x.borrow().clone();
}
```

Comment (Centri3) [reply tests/ui/needless_return.rs:371]: Yeah that does the same thing :+1: missed that before

**Comment by Centri3** [comment]
Thank you! @bors r+

**Pull request merged**